### DISCUSSION TITLE
返回的范围请求(HTTP range requests)视频流有很大几率会抛出异常

### DISCUSSION THREAD

**Discussion by ckvv** [midwayjs/midway#3688] [Q&A]
### Describe the problem（描述问题）

最小复现: https://github.com/ckvv/midway-range-request/blob/main/src/controller/home.controller.ts

启动后打开http://127.0.0.1:7001/ 有概率抛出以下异常信息，但是不影响视频播放
![截屏2024-03-05 17 54 40](https://github.com/midwayjs/midway/assets/30174970/f6642142-ff71-4361-bcf3-bdee6ab24fd9)

如果不用midwayjs实现的话就没有这个问题
https://github.com/ckvv/midway-range-request/blob/main/test.js

### Midway Versions（Midway 版本）

>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(not installed)
✓ @midwayjs/axios(not installed)
✓ @midwayjs/bootstrap(3.15.0)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/cache-manager(not installed)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.15.0)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(not installed)
✓ @midwayjs/cross-domain(not installed)
✓ @midwayjs/decorator(3.15.0)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✓ @midwayjs/i18n(not installed)
✓ @midwayjs/info(3.15.1)
✓ @midwayjs/jwt(not installed)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.15.0)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(not installed)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(not installed)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(not installed)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/tags(not installed)
✓ @midwayjs/typegoose(not installed)
✓ @midwayjs/typeorm(not installed)
✓ @midwayjs/upload(not installed)
✓ @midwayjs/validate(3.15.1)
✓ @midwayjs/version(not installed)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.15.1)
✓ @midwayjs/ws(not installed)
**********************************************************************
>> Check complete, all versions are healthy.
**********************************************************************

**Comment by czy88840616** [comment]
这个更像是对方主动关闭了连接，试试直接 pipe 到 ctx.res 吧

**Reply by ckvv** [reply]
报错了
```js
  return createReadStream(filePath).pipe(ctx.res);
```
```
2024-03-06 12:33:47.924 ERROR 84413 [-/::ffff:127.0.0.1/-/30ms GET /] Error: Cannot pipe, not readable
    at ServerResponse.pipe (node:_http_outgoing:1172:22)
    at respond (/Users/chenkai/Desktop/midway-range-request/node_modules/.pnpm/koa@2.15.0/node_modules/koa/lib/application.js:303:43)
    at handleResponse (/Users/chenkai/Desktop/midway-range-request/node_modules/.pnpm/koa@2.15.0/node_modules/koa/lib/application.js:184:34) {
  code: 'ERR_STREAM_CANNOT_PIPE'
}
```

**Reply by ckvv** [reply]
改成下面的不会再报错了，还是存在上面的`Error: write EPIPE` `Error: read ECONNRESET` 错误
 ```
ctx.respond = false;
return createReadStream(filePath).pipe(ctx.res);
```

**Comment by ckvv** [comment]
相关issue
https://github.com/koajs/koa/issues/1089
https://github.com/koajs/koa/issues/1345
